### PR TITLE
Support sharing of initializers between session via the Python API

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: Python API addition for the native/managed API introduced in https://github.com/microsoft/onnxruntime/pull/5092

**Motivation and Context**
Keep gap between native APIs and Python APIs as small as possible

### CODE REVIEW

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 1]
```diff
@@ -1849,60 +1849,62 @@ private void TestWeightSharingBetweenSessions()
             var dims = new long[] { 3, 2 };
             var dataBuffer = new float[] { 1.0F, 2.0F, 3.0F, 4.0F, 5.0F, 6.0F };
             var allocator = OrtAllocator.DefaultInstance;
-            var ortAllocationInput = allocator.Allocate((uint)dataBuffer.Length * sizeof(float));
-            unsafe
+            using (var ortAllocationInput = allocator.Allocate((uint)dataBuffer.Length * sizeof(float)))
```
Comment (hariharans29) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1852]: put `ortAllocationInput` in a using block so that it is disposed of - the rest are just formatting changes

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 2]
```diff
@@ -949,11 +949,12 @@ TEST(CApiTest, TestSharingOfInitializer) {
...
   // prepare expected inputs and outputs
   std::vector<int64_t> expected_dims_y = {3, 2};
-  std::vector<float> expected_values_y = {1.0f, 4.0f, 9.0f, 16.0f, 25.0f, 36.0f};
+  std::vector<float> expected_values_y = {2.0f, 2.0f, 12.0f, 12.0f, 30.0f, 30.0f};
 
   Ort::SessionOptions session_options;
   Ort::MemoryInfo mem_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
-  float data[] = {1., 2., 3., 4., 5., 6.};
+  // These values are different from the actual initializer values in the model
+  float data[] = {2., 1., 4., 3., 6., 5.};
```
Comment (hariharans29) [inline onnxruntime/test/shared_lib/test_inference.cc:957]: Little nit-picky but it is better to keep the manually provided overridden initializer different from the actual initializer in the model (the whole point of this API is to check if the manually provided initializer is being honored)

Comment (yufenglee) [reply onnxruntime/test/shared_lib/test_inference.cc:957]: good catch

---
In reply to: [501426123](https://github.com/microsoft/onnxruntime/pull/5407#discussion_r501426123) [](ancestors = 501426123)

File: onnxruntime/python/onnxruntime_pybind_state_common.h [hunk 3]
```diff
@@ -64,7 +64,7 @@ struct PyInferenceSession {
 
   void AddCustomOpLibraries(const std::vector<std::shared_ptr<CustomOpLibrary>>& custom_op_libraries) {
     if (!custom_op_libraries.empty()) {
-      custom_op_libraries_.reserve(custom_op_libraries_.size() + custom_op_libraries.size());
+      custom_op_libraries_.reserve(custom_op_libraries.size());
```
Comment (yufenglee) [inline onnxruntime/python/onnxruntime_pybind_state_common.h:67]: why this change?

Comment (hariharans29) [reply onnxruntime/python/onnxruntime_pybind_state_common.h:67]: it is not needed- minor cleanup from my original change.

File: onnxruntime/python/onnxruntime_pybind_mlvalue.cc [hunk 4]
```diff
@@ -656,12 +656,12 @@ void CreateGenericMLValue(const onnxruntime::InputDefList* input_def_list, const
...
 #endif
 
-  } else if (!accept_only_numpy_array && strcmp(Py_TYPE(value.ptr())->tp_name, "OrtValue") == 0) {
+  } else if (!accept_only_numpy_array && strcmp(Py_TYPE(value.ptr())->tp_name, PYTHON_ORTVALUE_OBJECT_NAME) == 0) {
     // This is an OrtValue coming in directly from Python, so assign the underlying native OrtValue handle
     // to the OrtValue object that we are going to use for Run().
     // This should just increase the ref counts of the underlying shared_ptrs in the native OrtValue
     // and the ref count will be decreased when the OrtValue used for Run() is destroyed upon exit.
-    *p_mlvalue = value.attr("_ortvalue").cast<OrtValue>();
+    *p_mlvalue = *value.attr(PYTHON_ORTVALUE_NATIVE_OBJECT_ATTR).cast<OrtValue*>();
```
Comment (hariharans29) [inline onnxruntime/python/onnxruntime_pybind_mlvalue.cc:664]: minor optimization: I think this should avoid creating a local temporary OrtValue instance whose contents are then copied into the feed OrtValue and then the temporary will be destroyed. With this change, the "raw" OrtValue instance in the Python object will be copied into the feed thus avoiding the local object creation

File: onnxruntime/python/onnxruntime_pybind_mlvalue.h [hunk 5]
```diff
@@ -21,6 +21,9 @@ namespace python {
 
 namespace py = pybind11;
 
+static constexpr const char* PYTHON_ORTVALUE_OBJECT_NAME = "OrtValue";
+static constexpr const char* PYTHON_ORTVALUE_NATIVE_OBJECT_ATTR = "_ortvalue";
+
```
Comment (yuslepukhin) [inline onnxruntime/python/onnxruntime_pybind_mlvalue.h:26]: Declaring statics in the header is not a good practice. Can you make it extern and declare it in the .cc file?

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 6]
```diff
@@ -1849,60 +1849,62 @@ private void TestWeightSharingBetweenSessions()
...
             {
-                float* p = (float*)ortAllocationInput.DangerousGetHandle();
-                for (int i = 0; i < dataBuffer.Length; ++i)
+                unsafe
                 {
-                    *p++ = dataBuffer[i];
+                    float* p = (float*)ortAllocationInput.DangerousGetHandle();
```
Comment (yuslepukhin) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1856]: >DangerousGetHandle [](start = 58, length = 18)

Where did this come from? Why not Pointer?

Comment (hariharans29) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1856]: This is the code from the original PR.

File: onnxruntime/python/onnxruntime_pybind_mlvalue.h (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.h:26]: Sure

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs (hunk 6 above)
Comment (pranavsharma) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1856]: Pointer seems to be marked internal and it's doing the same thing as [DangerousGetHandle](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle.dangerousgethandle?view=netcore-3.1#System_Runtime_InteropServices_SafeHandle_DangerousGetHandle).

File: onnxruntime/python/onnxruntime_pybind_mlvalue.h (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.h:26]: Hmm - I pushed a commit but it doesn't reflect here in the PR - it reflects in the branch when I look at its commits though

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs (hunk 6 above)
Comment (yuslepukhin) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1856]: I am afraid, it is not a good example. Can we make this test in the same manner as this: https://github.com/microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.Tests/OrtIoBindingAllocationTest.cs

Comment (yuslepukhin) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1856]: There are two reasons.  1) DangerouseGetHandle() is inherited from Safehandle and it is not a part of the official API we support. 2) There is a reason Pointer is internal and example of using Dangerous*() is allowing pointer leak outside.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 7]
```diff
@@ -1849,60 +1849,62 @@ private void TestWeightSharingBetweenSessions()
...
+                    for (int i = 0; i < dataBuffer.Length; ++i)
+                    {
+                        *p++ = dataBuffer[i];
+                    }
                 }
-            }
-            var dataBufferNumBytes = (uint)dataBuffer.Length * sizeof(float);
-            var sharedInitializer = OrtValue.CreateTensorValueWithData(ortCpuMemInfo, Tensors.TensorElementType.Float,
-            dims, ortAllocationInput.DangerousGetHandle(), dataBufferNumBytes);
+                var dataBufferNumBytes = (uint)dataBuffer.Length * sizeof(float);
```
Comment (yuslepukhin) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1862]: We do not need to use OrtAllocation here. Just create an array, pin memory and that would be it

**Review by pranavsharma — APPROVED** [review]

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs (hunk 6 above)
Comment (hariharans29) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs]: Resolved

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs (hunk 7 above)
Comment (hariharans29) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1863]: Fixed

**Review by yuslepukhin — APPROVED** [review]
:shipit:
